src/subscription/__tests__/subscribe-test.js

```diff
@@ -0,0 +1,215 @@
+        return eventEmitterAsyncIterator(pubsub, 'importantEmail');
+      }
+    };
+
+    function sendImportantEmail(newEmail) {
+      data.inbox.emails.push(newEmail);
+      // Returns true if the event was consumed by a subscriber.
+      return pubsub.emit('importantEmail', {
+        importantEmail: {
+          email: newEmail,
+          inbox: data.inbox,
+        }
+      });
+    }
+
+    const ast = parse(`
+      subscription ($priority: Int = 0) {
+        importantEmail(priority: $priority) {
+          email {
+            from
+            subject
+          }
+          inbox {
+            unread
+            total
+          }
+        }
+      }
+    `);
+
+    // GraphQL `subscribe` has the same call signature as `execute`, but returns
+    // AsyncIterator instead of Promise.
+    const subscription = subscribe(
+      emailSchema,
+      ast,
+      data,
+      null, // context
+      { priority: 1 }
+    );
+
+    // Wait for the next subscription payload.
+    const payload = subscription.next();
+
+    // A new email arrives!
+    expect(sendImportantEmail({
+      from: 'yuzhi@graphql.org',
+      subject: 'Alright',
+      message: 'Tests are good',
+      unread: true,
+    })).to.equal(true);
+
+    // The previously waited on payload now has a value.
+    expect(await payload).to.deep.equal({
+      done: false,
+      value: {
+        data: {
+          importantEmail: {
+            email: {
+              from: 'yuzhi@graphql.org',
+              subject: 'Alright',
+            },
+            inbox: {
+              unread: 1,
+              total: 2,
+            },
+          },
+        },
+      },
+    });
+
+    // Another new email arrives, before subscription.next() is called.
+    expect(sendImportantEmail({
+      from: 'hyo@graphql.org',
+      subject: 'Tools',
+      message: 'I <3 making things',
+      unread: true,
+    })).to.equal(true);
+
+    // The next waited on payload will have a value.
+    expect(await subscription.next()).to.deep.equal({
+      done: false,
+      value: {
+        data: {
+          importantEmail: {
+            email: {
+              from: 'hyo@graphql.org',
+              subject: 'Tools',
+            },
+            inbox: {
+              unread: 2,
+              total: 3,
+            },
+          },
+        },
+      },
+    });
+
+    // The client decides to disconnect.
+    expect(await subscription.return()).to.deep.equal({
+      done: true,
+      value: undefined,
+    });
+
+    // Which may result in disconnecting upstream services as well.
+    expect(sendImportantEmail({
+      from: 'adam@graphql.org',
+      subject: 'Important',
+      message: 'Read me please',
+      unread: true,
+    })).to.equal(false); // No more listeners.
+
+    // Awaiting a subscription after closing it results in completed results.
+    expect(await subscription.next()).to.deep.equal({
+      done: true,
+      value: undefined,
+    });
+  });
+
```

Comment:
For most subscriptions, the AsyncIterable stream is infinite (take for example, likes on a post, there's no theoretical "final like"). Rather, we need to fluently express the case of unsubscribing.